### PR TITLE
DM-28530: Add ready_to_take_data functionality in ts_observatory_control.

### PR DESCRIPTION
**Pull request by tribeiro**
* Add new feature to support synchronization between BaseTCS and BaseCamera.
* Implement synchronization feature in ATCS.
* Implement placeholder for synchronization feature in MTCS.
* Added documentation on how to use the synchronization between TCS and Camera.

### CODE REVIEW

File: python/lsst/ts/observatory/control/auxtel/atcs.py [hunk 1]
```diff
@@ -1524,6 +1524,140 @@ def unset_azel_slew_checks(self, checks):
         self.check.adome = checks.dome
         self.check.atdometrajectory = checks.dometrajectory
 
+    async def _ready_to_take_data(self):
+        """Wait until ATCS is ready to take data.
+
+        If cancelled or some condition times out or raise an exeception it will
```
Comment (wvreeven) [inline python/lsst/ts/observatory/control/auxtel/atcs.py:1530]: Change `raise` to `raises`.

File: python/lsst/ts/observatory/control/auxtel/atcs.py [hunk 2]
```diff
@@ -1524,6 +1524,140 @@ def unset_azel_slew_checks(self, checks):
...
+        """
+        # Things to check
+        # 1 - ATMCS evt_allAxesInPosition.inPosition == True
+        # 2 - All ATAOS corrections are applied
+        # 3 - ATMCS and ATDome positions are within specified range.
+        ready = False
+        try:
+            while not ready:
+                # This loop will run until all conditions are met. Note that it
+                # may happen that one condition will be me when we run it the
```
Comment (wvreeven) [inline python/lsst/ts/observatory/control/auxtel/atcs.py:1542]: Change `me` to `met`.

File: python/lsst/ts/observatory/control/auxtel/atcs.py [hunk 3]
```diff
@@ -1524,6 +1524,140 @@ def unset_azel_slew_checks(self, checks):
...
+                self.log.debug(
+                    f"Ready to take data:: atmcs={check[0]}, ataos={check[1]}, atdome={check[2]}."
+                )
+                ready = all(check)
+        except asyncio.CancelledError as e:
+            self._ready_to_take_data_future.set_exception(e)
+        except asyncio.TimeoutError as e:
+            self._ready_to_take_data_future.set_exception(e)
+        except Exception as e:
+            self._ready_to_take_data_future.set_exception(e)
```
Comment (wvreeven) [inline python/lsst/ts/observatory/control/auxtel/atcs.py:1558]: Is it necessary to split out these different errors like this? In the end they all call the same method so lines 1553 to 1556 are superfluous.

File: python/lsst/ts/observatory/control/auxtel/atcs.py [hunk 4]
```diff
@@ -1524,6 +1524,140 @@ def unset_azel_slew_checks(self, checks):
...
+        ret_val = await asyncio.gather(
+            self.rem.ataos.evt_atspectrographCorrectionStarted.next(
+                flush=True, timeout=self.fast_timeout
+            ),
+            self.rem.ataos.evt_atspectrographCorrectionCompleted.next(
+                flush=True, timeout=self.fast_timeout
+            ),
+            return_exceptions=True,
+        )
+        # If they are both execeptions it probably means the ATAOS got into a
```
Comment (wvreeven) [inline python/lsst/ts/observatory/control/auxtel/atcs.py:1598]: Change `they are both` to `they both raise`.

File: python/lsst/ts/observatory/control/auxtel/atcs.py [hunk 5]
```diff
@@ -1524,6 +1524,140 @@ def unset_azel_slew_checks(self, checks):
...
+                if not isinstance(corr, asyncio.TimeoutError)
+            ]
+
+            self.log.debug(f"Received corrections: {correction_received}.")
+            return False
+
+    async def atmcs_in_position(self):
+        """Check if atmcs is in position.
+
+        This method will try to get the next even published after the call. If
```
Comment (wvreeven) [inline python/lsst/ts/observatory/control/auxtel/atcs.py:1640]: Change `even` to `event`.

File: python/lsst/ts/observatory/control/auxtel/atcs.py [hunk 6]
```diff
@@ -1524,6 +1524,140 @@ def unset_azel_slew_checks(self, checks):
...
+            # This means we got some data. We are not in a stable state yet.
+            correction_name = ["started", "completed"]
+            correction_received = [
+                name
+                for name, corr in zip(correction_name, ret_val)
+                if not isinstance(corr, asyncio.TimeoutError)
+            ]
+
+            self.log.debug(f"Received corrections: {correction_received}.")
+            return False
```
Comment (wvreeven) [inline python/lsst/ts/observatory/control/auxtel/atcs.py:1635]: My main concern with this approach is that `ret_val` may simply have retrieved the latest CorrectionStarted and CorrectionCompleted events so technically speaking the corrections have been applied but the code forces to wait for the next invocation where those two events time out. This seems like unnecessarily waiting for a timeout to happen. If `ret_val` contains data in the outer else block then applying the corrections may have finished if `private_sndStamp` of CorrectionStarted < `private_sndStamp` of CorrectionCompleted, right?

File: python/lsst/ts/observatory/control/base_tcs.py [hunk 7]
```diff
@@ -952,6 +955,20 @@ async def _offset(self, offset_cmd):
...
+        to determine the condition of the system.
+        """
+        if self._ready_to_take_data_future.done():
+            self._ready_to_take_data_future = asyncio.Future()
+            self._ready_to_take_data_task = asyncio.create_task(
+                self._ready_to_take_data()
+            )
+            return self._ready_to_take_data_future
+        else:
+            return self._ready_to_take_data_future
```
Comment (wvreeven) [inline python/lsst/ts/observatory/control/base_tcs.py:970]: You can get rid of the `else` block and simply return `self._ready_to_take_data_future` in a line at the same indentation as the `if` statement.

File: python/lsst/ts/observatory/control/mock/atcs_mock.py [hunk 8]
```diff
@@ -210,6 +223,31 @@ async def atptg_telemetry(self):
 
             await asyncio.sleep(1.0)
 
+    async def ataos_telemetry(self):
+
+        self.ataos.evt_correctionEnabled.set_put()
+
+        correction_times = cycle((1, 1, 2, 4, 10))
+        while self.run_telemetry_loop:
+            if self.ataos.evt_summaryState.data.summaryState == salobj.State.ENABLED:
```
Comment (wvreeven) [inline python/lsst/ts/observatory/control/mock/atcs_mock.py:232]: Not entirely sure, but shouldn't this `if` be outside the `while` loop?

**Review by wvreeven — CHANGES_REQUESTED** [review]
One concern is the way the code is forcing for a timeout while this appears to be unnecessary so please have a look at that. For the rest some minor typos and a few other possible improvements.

File: python/lsst/ts/observatory/control/auxtel/atcs.py (hunk 6 above)
Comment (tribeiro) [reply python/lsst/ts/observatory/control/auxtel/atcs.py:1635]: so, the problem is that, if a correction was completed you cannot guarantee that there won't be another one coming just after it. The only way to be more certain about it is to make sure all corrections scheduled where completed and no new correction is done in that grace period of time.

File: python/lsst/ts/observatory/control/mock/atcs_mock.py (hunk 8 above)
Comment (tribeiro) [reply python/lsst/ts/observatory/control/mock/atcs_mock.py:232]: no, this loop is constantly running, even of the state is not enabled.

File: python/lsst/ts/observatory/control/auxtel/atcs.py (hunk 3 above)
Comment (tribeiro) [reply python/lsst/ts/observatory/control/auxtel/atcs.py:1558]: yes, the problem is that some exceptions are no longer children of `Exception`, like `CancelledError`. So I am future-proofing the code if they ever decide to make the same with `TimeoutError`.

File: python/lsst/ts/observatory/control/base_tcs.py (hunk 7 above)
Comment (tribeiro) [reply python/lsst/ts/observatory/control/base_tcs.py:970]: good catch

**Comment by tribeiro** [comment]
I am not particularly happy either.. but because the distributed nature of the system there is no good way of knowing how many corrections the ATAOS will do to the telescope position. It may be 1, 2 or more. What we know is that we will have a burst of corrections when we setup the instrument and then, they will stop, until you change the instrument setup again. These usually happens back to back at a ~1Hz frequency. So, 5s of timeout is the best compromise I could find.

File: python/lsst/ts/observatory/control/auxtel/atcs.py (hunk 3 above)
Comment (wvreeven) [reply python/lsst/ts/observatory/control/auxtel/atcs.py:1558]: OK, good idea.

File: python/lsst/ts/observatory/control/auxtel/atcs.py (hunk 6 above)
Comment (wvreeven) [reply python/lsst/ts/observatory/control/auxtel/atcs.py:1635]: OK, thanks for the explanation.

File: python/lsst/ts/observatory/control/mock/atcs_mock.py (hunk 8 above)
Comment (wvreeven) [reply python/lsst/ts/observatory/control/mock/atcs_mock.py:232]: Yeah I figured which is why I wasn't sure. Thanks for the explanation.

**Review by wvreeven — APPROVED** [review]
Looks good to me!

**Pull request merged**